### ISSUE TITLE
Can you bring this to PHP RFC? 

### ISSUE THREAD

**Issue by HallofFamer** [nikic/scalar_objects#20]
I took a look at it and I very like what is being implemented. In my own framework I use classes for Int, Float, String and Array, but sometimes it can be a bit tedious(like having to write new String("abc") than just "abc"). This library will fix a lot of issues in my codebase, and make coding 10x more enjoyable. Also if its implemented at the PHP core, there's a chance to suggest a way to make even more elegant syntax work, such as "abc"->toUpper() which does not happen yet. So, will you consider moving this to RFC and possibly have PHP internals implement it in PHP 7?

**Comment by terah** [comment]
I think there are a lot of PHP devs that would love to see this in the core...

**Comment by HallofFamer** [comment]
yeah of course, for developers who want to use strongly typed PHP scalars, this will help tremendously. Please do consider bringing it to PHP RFC.

**Comment by nikic** [comment]
I won't have the time to propose this for inclusion in PHP 7. Adding this to core would require quite a lot of work and feature freeze is not far away.

By the way, I already added support for syntax like `"abc"->toUpper()` and `[1, 2, 3]->reverse()` to PHP 7 (it will just throw a runtime error). So once this extension is ported to PHP 7, you'll be able to use that syntax in conjunction with the ext.

**Comment by HallofFamer** [comment]
I see, thats a pity, I really hope to see this on PHP 7.0, but guess it wont be there at least till 7.1... Still, you can start an RFC to prepare for the future, even if it will not be going to voting phase anytime soon, right?

And its very good news, I very love the syntax like "mystring"->length() and [2, 5, 4, 6]->sort(). It will be very useful, cant wait for the new Scalar object for PHP 7!

**Comment by TomBZombie** [comment]
Shame, I'd have also loved to have seen this in PHP7. As I mentioned in one of the other issues, this is a perfect opportunity to fix the inconsistent function naming and parameter orders... $str->replace(), $array->combine() etc.

**Comment by alanwillms** [comment]
@nikic what about modifying `$self` within the handlers? It will be possible in PHP 7?

**Comment by nikic** [comment]
@alanwillms Nope, the current VM implementation doesn't allow this. With enough effort it might be possible, but would probably have negative performance impact for all other method calls (which is pretty much a no-go).

**Comment by alanwillms** [comment]
@nikic not modifying `$self` may be a good thing since it enforces immutability.

Now that PHP 7.0 is frozen and won't have any new features, will you open a RFC targeting 7.1?

**Comment by c9s** [comment]
+1

**Comment by Strate** [comment]
:+1: for this in php core

**Comment by TheXardas** [comment]
Yep, +1.

**Comment by alexkvak** [comment]
+1

**Comment by SirSnyder** [comment]
+1

**Comment by kandy** [comment]
Also will be good to split to two RFC. One about engine level ability. And one about user space registration.

**Comment by dominikzogg** [comment]
+1

**Comment by robob4him** [comment]
@nikic "once this extension is ported to PHP 7"

Any progress on this?

**Comment by vbsessa** [comment]
+1

**Comment by mbrzuchalski** [comment]
:+1:

**Comment by dgafka** [comment]
+1

**Comment by sjaworski** [comment]
+1

**Comment by jkonefal** [comment]
:+1:

**Comment by event15** [comment]
+1

**Comment by KennetPL** [comment]
+1

**Comment by danizord** [comment]
:+1:

**Comment by levan1993** [comment]
+1

**Comment by ctrl-f5** [comment]
+1

**Comment by marcelgsantos** [comment]
:+1:

**Comment by brzuchal** [comment]
@nikic are there any chances to bring this to RFC into the core of PHP? This feature is awesome and there are plenty of people looking forward for this.

**Comment by devypt** [comment]
+1

**Comment by nilportugues** [comment]
+1

**Comment by ciscoo** [comment]
Just a reminder everyone that you can add a reaction to comments rather than cluttering the comments with +1.

**Comment by kalvisbuls** [comment]
:+1:

**Comment by whatsupdan** [comment]
+1

**Comment by SerafimArts** [comment]
+1

**Comment by SCIF** [comment]
+1

**Comment by eduardor2k** [comment]
+1

**Comment by prwnr** [comment]
Go for it! +1

**Comment by bordeux** [comment]
+15

**Comment by exts** [comment]
+1

**Comment by ctrl-f5** [comment]
**please stop replying "+1"**, to vote just add reaction to the original comment.
You are spamming all participants' mailboxes.

**Comment by jens1o** [comment]
Will this be possible in PHP 7.2?

**Comment by ctrl-f5** [comment]
no, there hasn't even been an RFC for this, so it can not possibly be on scheduled to be in a future release.

**Comment by jclyons52** [comment]
What can we do to bring this closer to an RFC?

**Comment by Rastusik** [comment]
+1

**Comment by mustafatoker** [comment]
+1

**Comment by dennisoehme** [comment]
+1

**Comment by mpyw** [comment]
Hope for PHP 8.x or 9.x

**Comment by qwerty199369** [comment]
This feature is awesome! 👍  Please do consider bringing it to PHP RFC. @laruence @nikic 👍

**Comment by alanwillms** [comment]
PHP core developers are discussing about tagging the next release after 7.3 as 8.0. Maybe now is the time.

**Comment by q2dg** [comment]
An entire year has passed already...do you know any news on this? Thanks!

**Comment by mpyw** [comment]
So interesting but handlers should be applied in a specific scope.

**Comment by HRADigital** [comment]
> An entire year has passed already...do you know any news on this? Thanks!

It doesn't seam to be planned for v8!

**Comment by mpyw** [comment]
[C# Extension Method](https://www.tutorialsteacher.com/csharp/csharp-extension-method)

C#'s Extension Method is definitively our goal.

**Comment by Atulin** [comment]
So it hasn't gone to an RFC for... four years now? What's stopping it?

**Comment by HallofFamer** [comment]
@mpyw yup this one has the potential to allow any types to be extended instead of just primitive types, though unlike C# which is compiled, in PHP you will have to manually include the files that contain extension methods. 

Anyway, I think, its about time to bring this to PHP RFC. It will be a neat feature for PHP 8, a good chance to introduce OO API for scalar types, as well as fixing the messy function names that cannot be touched due to backward compatibility issue.

**Comment by vbsessa** [comment]
"fixing the messy function names that cannot be touched due to backward compatibility issue".

Great point.

**Comment by avierr** [comment]
+1000

**Comment by mvorisek** [comment]
> @alanwillms Nope, the current VM implementation doesn't allow this. With enough effort it might be possible, but would probably have negative performance impact for all other method calls (which is pretty much a no-go).

What is the limitation, why the scalar value can not be passed as a variable by reference?

**Comment by nikic** [comment]
@mvorisek Consider something like `$a[$b][$c]->push($d)`. We need to know that `$a[$b][$c]` is an array in order to determine that the call should be performed by-reference. However, we already need to convert `$a`, `$a[$b]` and `$a[$b][$c]` into references before we know about that. The only way to make this work is to *always* fetch by reference, which would both result in a semantic change for existing object accesses, and in a large performance regression.

Supporting something like `$a->push($b)` is simple, but supporting the general case is, I believe, semantically impossible.

**Comment by mvorisek** [comment]
@nikic `$a[$b][$c]->push($d)` is good example.

As `$a[$b][$c]` can be an instance (and the `push` can be a regular method then), PHP must still wait until the expression is executed and then it can pass the resulting element (`$a[$b][$c]`) by reference.

On the other side something like `1 ? $a[$b][$c]` : 0` (result from expression) should never be allowed to be modified by reference.

So this should be possible with methods that modified the original scalar object:
```
$a = [['x']];
$a->push('y');
$a[0]->push('y');
($a)->push('y'); // reset(($a)) works, so this should work too
(['x'])->push('y'); // it should work too, as the value is newly initialized and it is not a result from expression (i.e. not `1 ? ['x'] : 0`)
```

What do you think?

**Comment by mvorisek** [comment]
@nikic Also what do you think about a registration only for some namespace prefix(es)? This would widen the possibility of using the scalar objects in larger projects/packages a lot.

**Comment by brunoggdev** [comment]
Hi! Just found this repo and it looks amazing. Any chance it gets on a RFC now? Maybe for PHP 9?

**Comment by adrolter** [comment]
@brunoggdev See @nikic's last comment from 2019. I doubt much has changed in that regard.
